Refactor fulltext search functions and class Doku_Indexer

ssahara opened the pull request:
Working in progress with issue #2880, 
  * idx_* functions in inc/indexer.php are encapsulated into singleton Indexer class (former Doku_Indexer)
  * all static functions in inc/fulltext.php are encapsulated into FulltextSearch class 
  * two methods in FulltextSearch (backlinks and mediause), those lookup metadata index and not related to full-text search, are separated into MetaSearch class
  * introduce new namespace "dokuwiki\Search", into which Indexer, FulltextSearch and MetaSearch classes are moved.

For the moment, all public functions in fulltext and indexer are modified to use relevant class methods in Search namespace and marked as "deprecated", but left for compatibility. I am sticking to change only fulltext.php and indexer.php files in this work. Further development must be needed.

I hope that I understand the issue #2880 correctly, but I wonder these changes could get in goods with core development teams members?

Here are some points to consider prior to go further developments: 
1. `MetaSearch::mediause()`, former `ft_mediause()`, could be defined in inc/media.php file as `getUsedMedia()` ?
2. `MetaSearch::backlinks()`, former `ft_backlinks()`, could be defined in inc/common.php file as `getBacklinks()` ?
3. If above two method can be moved to different places, class MetaSearch can be removed. Then, The class name "FulltextSearch" can be renamed to "Search".
4. New namespace "Fulltext" might be better that contains Indexer and (Fulltext)Search classes?
5. global constant `FT_SNIPPET_NUMBER` : where should be defined after when inc/fulltext.php file will be removed?
7. Should we use namespace constants `INDEXER_VERSION` and `IDX_MINWORDLENGTH` instead of global constatnts?

splitbrain commented:
First of all: thanks a lot! This is greatly appreciated. I currently don't have much time to look at this in detail. But let me see if I can answer your questions:

1. I would keep it as `MetaSearch::mediause()` because the stuff in `media.php` is ripe for refactoring as well (in general, I'd like to get rid of most of the global function stuff)
2. Same as 1. - keep it as `MetaSearch::backlinks()`
3. not sure yet, see below
4. I think `Index` might make the most sense for the namespace name, but see below
5. I guess it should not be a constant anymore, but a member of the appropriate class (with getter/setter)
6. same as 5, make them class members

Some things to think about:

* we have an index for pages with
  * words, sorted by length (w*.idx)
  * a list of pages (page.idx)
  * a word-page relation, sorted by word length (i*.idx)
  * a reverse lookup of page-word relations (pageword.idx)
* we have an index for metadata
  * a list of pages (page.idx)
  * a list of titles matching the pages (title.idx)
  * multiple subindexes (currently for media and references) that work similar to the page index (*_w.idx, *_p.idx, *_i.idx)
* we have a query parser that
  * handles search syntax and transforms it into an array
  * that array is used to search the (page)index
* we have simpler access mechanisms to the index data, for example
  * mediainuse()
  * backlinks()
  * i think there are a few more

There are things that are similar, but not the same. There is indexing and querying functionality.

What I would like to see in the end, is a class hierarchy where we have the core functionality of an index at the base and specialized children to implement a "PageWordIndex" or a "MediaRelationIndex". 

The Index should provide some way to retrieve it's data and Query classes should be able to work with different indexes.

It should be simple for a plugin (say the tag plugin or the docsearch plugin) to create it's own, new index to store and quickly access data in it by simply inheriting some "BaseIndex".

I hope that makes sense?

ssahara commented:
Thanks for your response and advice.
* The Indexer class has now devided into MetadataIndex, PagewordIndex, and PageIndex, these three extend an AbstractIndex class that provide basic methods to access *.idx files.
* Fulltext Search (ft_*) functions are separated to and encapsulated in QueryParser and FulltextSearch classes. The page name search methods ft_pageLookup() etc. are moved to MetadataSerach class.
* QueryParser class created.
* Tokenizer class separated from PagewordIndex class.

 I hope these structualization of fulltext search engine could get in goods with core development teams members?

Some ToDOs:
1. `Event::trigger()` needs to support static methods as event handler. This is not directly concerned with fulltext search feature, however some fix must need to allow static methods event handlers only for legacy PHP 5.6 environment.
2. fix inc/load.php file, loading fulltext.php and indexer.php is not necessary.
3. I wonder PageIndex.php should be renamed to Indexer.php, which is better?

micgro42 commented:
@ssahara This pull request has still WIP in its title, is it ready for review?

ssahara commented:
@micgro42 Sorry I could not enough time to complete this PR, hopefully it will take more one 1 week.

micgro42 commented:
@ssahara Don't worry and take all the time you need. I was just unsure whether you already wanted a detailed review or whether you are still working on it.

I had a cursory look and I like the general direction that this is going 👍
Some observations:

- Since the new index classes are intended to be singletons, it might be possible to turn many of the new static methods into instance methods. That would make it nicer to use these classes with dependency injection in the future. That, in turn, would make testing much easier.
- `dbg_deprecated()` takes a hint to the alternative as its first argument. It gets the deprecated function from the backtrace.

ssahara commented:
@micgro42 Thanks for your advice. As you suggested, it is possible to convert static methods in MetadataSearch, FulltextSearch to instance methods. Also I have fixed wrong usage of `dbg_deprecated()` in inc/deprecated.php file. 
I have removed "WIP" from this PR title, which means ready for review.

ssahara commented on inc/Search/Indexer.php:
```diff
@@ -861,50 +318,52 @@ public function getPages($key=null) {
-            $min = 1;
-        if ($max < $min)
-            $max = 0;
+    public function histogram($key = null, $min = 1, $max = 0, $minlen = 3)
```
The `histogram()` seems to be used for fulltext and metadata index, therefore the 1st parameter should be $key specifies the target index. Other params are not relevant with index type.
Anyway, `histogram()` is not used anywhere but used only in indexer_histogram.test.php file. The test covers metadata index only.

micgro42 replied:
I wondered where it is used as well, and it turns out there seems to be a single, but somewhat maintained plugin that uses it: the tagfilter plugin: https://codesearch.dokuwiki.org/search?project=dokuwiki&project=plugin&project=template&q=&defs=&refs=histogram&path=&hist=&type=

While your motivation makes sense to me, changing the signature of the `histogram` method breaks that plugin's "compare tags" feature.

In general, it seems the tagging plugins use the index a lot and looking into them highlighted a few more problems. That tagfilter plugin actually calls `require_once(DOKU_INC.'inc/indexer.php');`. That is not needed anymore and I just made a PR there to remove that: lisps/tagfilter#15

But then the tag plugin failed with the following error:

```
Fatal error: Uncaught Error: Call to undefined method dokuwiki\Search\Indexer::lookupKey() in /home/michael/public_html/dokuwiki/lib/plugins/tag/helper.php on line 353
```

I think this is valid and that `Indexer` class should probably have a deprecated method pointing to the correct one on `MetaIndexer`

ssahara replied:
I see. I will check [tag plugin](https://github.com/dokufreaks/plugin-tag/blob/58b5adf7fd8cb440aabce4a141806a3a931a8254/helper.php#L352-L353
).  Just idea, should idx_get_index() return MetadataIndex?
```` php
        $indexer = idx_get_indexer();
        $pages = $indexer->lookupKey('subject', $clean_tags, array($this, '_tagCompare'));
````

ssahara replied:
changes in `histogram()` have reverted. 5f9bd52
and deprecated `idx_getindexer()` now returns MetadataIndex b4c31a2, this make the tag plugin work fine with this branch!

ssahara replied:
I made a PR to tag plugin to fix PHP74 issue
https://github.com/dokufreaks/plugin-tag/pull/205

ssahara commented on inc/search.php:
```diff
@@ -22,22 +22,23 @@
  *                          'date' to sort by filemtime; leave empty to skip sorting.
```
Yes. it have white space changes only. In this PR, "Search" namespace will created, and I just checked (with adjusting whitespaces) whether the inc/search.php file should be converted a class in Search namespace. I think in future when functions in inc/search.php will be converted to class methods, the class will be placed other namespace than Search.

ssahara commented on _test/tests/inc/remoteapicore.test.php:
```diff
@@ -85,11 +92,13 @@ public function test_getPageList() {
         $file = wikiFN($id);
 
-        idx_addPage($id); //full text search depends on index
+        $Indexer = Indexer::getInstance(); //full text search depends on index
```
Agree. The comment removed.

ssahara commented on inc/Search/Indexer.php:
```diff
@@ -3,852 +3,309 @@
+        } else {
+            // update search index
+            $result = $this->addPage($page, $verbose, $force);
+            return $result;
```
fixed 11d2e7d

ssahara commented on inc/Search/FulltextIndex.php:
```diff
@@ -0,0 +1,518 @@
+     *
+     * @param string    $page   a page name
+     * @param string    $text   the body of the page
+     * @param bool      $requireLock
```
Done. $requireLock should be false only if the caller methods is responsible for data/index directory lock, instead of this method.

Klap-in commented:
Hi, there is a conflict that needs to be solved. 
Are there other points of attention?

ssahara commented:
The conflict has now resolved.
I appreciate it if someone could help to fix scrutinizer claims?

moisesbr-dw commented:
Hi.

If possible, it would be simpler to firstly merge #2037 , as the indexer issue here have changed the files A LOT.
#2037 is way simpler, a new file for internationalized sort functions and the corresponding changes in less than 20 files, *including fulltext.php.*

splitbrain commented on inc/Search/AbstractIndex.php:
```diff
@@ -0,0 +1,403 @@
+        // return PID when it is in the cache
+        // avoid expensive addIndexKey operation for the most recently
+        // requested pages by using a cache
+        if (isset(static::$pidCache[$page])) return static::$pidCache[$page];
```
I think making the page cache static might be wrong. This makes all child classes use the same single cache array but children might want to use a different "page". Eg. a media index might store media IDs instead of pages. Or am I missing something?

ssahara replied:
AFAIK DokuWIki does not have **media** index feature. We have FulltextIndex and MetadataIndex, both use only pageid as the static variable is named as `$pidCache`.

ssahara replied:
@splitbrain I had wrong understanding about how the pidCache work between Indexer and other index classes. 
The changes in a32da6d does make Indexer pass a pid to MetadataIndex and FulltextIndex, and these subclass does not need to call `getPID()` to get numeric page id to be indexed. It might also mean the pidCache is unnecessary?

splitbrain replied:
The idea behind having this abstract indexing class, is that we could have additional indexes in the future and there the PIDs might refer to something else, not the common page.idx but a media.idx or whatever. A static variable will be the same for all instances of the class and it's descendants. So in that case a static cache here would be incorrect. ..

splitbrain replied:
Ah I stand corrected. I experimented a bit and an inheriting class can actually redefine a static member, so it's cache is separate. TIL.

splitbrain commented:
@ssahara I was having a look at the scrutinizer issues and noticed that we're still raising user errors, which isn't good. So I  tried to replace them by exceptions in 15f699a I still think this could be improved. Can you have a look and tell me what you think about it?

ssahara commented:
@splitbrain , thanks for updating the code. I would like to share some findings during a way to find something to improve the code. 
1. Failures in `AbstractIndex::lock()/unlock()` are run-time errors and should be treated as **IndexLockException**.  We could declare them as void instead of returning true?
2.  It looks that **IndexWriteException** will happen in `AbstractIndex::saveIndex()/saveIndexKey()`
 when temporay file could not created. It is run-time errors and should be treated as exception. I think it is better to throw **IndexWriteException** in `saveIndex()/saveIndexKey()`. We also could declare them as void? 
3. I think **IndexAccessException** might be **logic error** rather than run-time error. I will further investigate if it is appropriate or not.

ssahara commented on inc/TaskRunner.php:
```diff
@@ -197,7 +197,12 @@ protected function runIndexer()
+            return $Indexer->dispatch($ID, true);
+        } catch(Search\Exception\SearchException $e) {
+            echo get_class($e) . ' : ' . $e->getMessage();
+            return false;
```
SearchException messages are also debug.log file using dbglog(). see 72ebc99

Klap-in commented:
What is the status of this PR? Is it ready for a merge (after resolving current conflicts)?
If it is done, then I think it would be nice to include it yet in coming Igor release. If there is still too much work to do, then it is maybe not so realistic idea, I could imagine that some time between the merge and release is useful for testing more plugins. (or having an release candidate, to encourage testing of plugins.)

ssahara commented:
@Klap-in , I have just resolved conflicts with current master branch. For the moment, I have no idea for further developments.

Klap-in commented:
@micgro42 and @splitbrain, are you happy with merging this PR?

micgro42 commented:
> @micgro42 and @splitbrain, are you happy with merging this PR?

Eh, I don't think I can possibly be happy with a PR that changes thousands of lines of code, and both significantly refactors code and changes behavior.

But I'm not blocking this. Merging it, crossing our fingers, and trying to fix the bugs as we find them, seems like the best way forward by now.

splitbrain commented on inc/Search/Indexer.php:
```diff
@@ -3,1214 +3,315 @@
-     * @var array $pidCache Cache for getPID()
-     */
-    protected $pidCache = array();
+class Indexer extends AbstractIndex
```
To me it makes no sense that this class inherits from AbstractIndex. This muddles the class hierarchy.

* AbstractIndex should define what makes an index (methods for adding, removing, etc)
* FulltextIndex and MetadataIndex should implement specifics inheriting from Abstract Index
* Indexer is not an index in itself. It's a worker that uses the above indexes.

splitbrain commented:
Hi @ssahara 

As @micgro42 mentioned, your pull request is huge and it is a major undertaking to review it. I spent the day to do so and haven't even seen everything. I don't blame you. Refactoring a complex thing like this is a massive undertaking and often results in massive PRs. 

Even though it is an improvement, I don't want to merge it in its current form. The main reason is that I think it is not going far enough. Looking through the code I noticed that many of the quirks we had in the old code transformed into the new code. 

My fear is, when I merge this now, nobody will ever look at it again and improve further. Instead I suggest the following.

I merged the code into the `searchIndex` branch here and I added you as a repo contributor. This makes it easier for your to collaborate with others here like @Klap-in, @micgro42 and me.

The next step is to improve the `searchIndex` branch by a series of smaller PRs that are easier to review and move the branch forward bit by bit.

I will start with a first PR that will only rearrange the contents of the AbstractIndex into smaller classes. I will also open a new PR based on the `searchIndex` branch to outline my idea of the basic architecture.

ssahara commented:
I agree this refactoring has not reached enough level to have worth to be merged, and a series of smaller PRs approach is preferable. When I started this work two years ago, I could not image any clear goal. I still have not clear successful image to implement better `searchindex`,  but It may be possible to contribute something based on the experience of this refactoring work. 
Many thanks to @micgro42 , @Klap-in and @splitbrain for your time to review many code changes.
